### PR TITLE
Refactor invalidation log inclusion

### LINKED ISSUE #3833: Refactor `tsl/src/continuous_aggs/invalidation.c:invalidation_add_entry`
The function `invalidation_add_entry` have too much responsibilities.

Let's refactor it separating the code for invalidating continuous
aggregates and invalidating associated hypertables.

### PR DESCRIPTION
**Pull request by fabriziomello**
Commit 97c2578ffa6b08f733a75381defefc176c91826b overcomplicated the
`invalidate_add_entry` API by adding parameters related to the remote
function call for multi-node on materialization hypertables.
    
Refactored it simplifying the function interface and adding a new
function to deal with materialization hypertables on multi-node
environment.

Fixes #3833

### CODE REVIEW

File: tsl/src/continuous_aggs/invalidation.c
Comment (mfundul) [inline tsl/src/continuous_aggs/invalidation.c:305]: The comment can't talk about both the "continuous aggregate invalidation log" and the "hypertable invalidation log"

File: tsl/src/continuous_aggs/invalidation.c
Comment (mfundul) [inline tsl/src/continuous_aggs/invalidation.c:308]: The name `invalidate_continuous_agg()` is confusing.

The terminology is:
- "hypertable invalidation log"
- "continuous aggregate invalidation log" or "materialization invalidation log"

File: tsl/src/continuous_aggs/invalidation.c
Comment (mfundul) [inline tsl/src/continuous_aggs/invalidation.c:330]: This does not invalidation a materialization hypertable at all. This is only about the underlying hypertable invalidation log. Again the comment talks both about "cagg invalidation log" and "hypertable invalidation log" which is wrong.

**Review by mfundul — CHANGES_REQUESTED** [review]
The terminology for function names and the comment descriptions need to be fixed.

File: tsl/src/continuous_aggs/create.c
Comment (gayyappan) [inline tsl/src/continuous_aggs/create.c:560]: why is this mat_ht now?

Comment (mfundul) [reply tsl/src/continuous_aggs/create.c:560]: mat_ht should mean materialization hypertable

Comment (gayyappan) [reply tsl/src/continuous_aggs/create.c:560]: Yes, we add a invalidation log entry for the original hypertable , not the materialization hypertable. So why is this changed to use mat_ht?

File: tsl/src/continuous_aggs/invalidation.c
Comment (gayyappan) [reply tsl/src/continuous_aggs/invalidation.c:308]: This rename is confusing/misleading. The function adds an entry to the invalidation log. 
In general, I find the refactoring/renaming changes pretty confusing.

File: tsl/src/continuous_aggs/invalidation.c
Comment (nikkhils) [inline tsl/src/continuous_aggs/invalidation.c:348]: Looks like this needs to be covered via tests.

**Review by nikkhils — APPROVED** [review]
LGTM. But need to may need to improve tests for code coverage as per comment

File: src/cross_module_fn.h
Comment (mfundul) [inline src/cross_module_fn.h:104]:
```suggestion
	void (*continuous_agg_invalidate_raw_ht)(const Hypertable *raw_ht, int64 start, int64 end);
	void (*continuous_agg_invalidate_mat_ht)(const Hypertable *raw_ht,
```

File: tsl/src/continuous_aggs/invalidation.c
Comment (mfundul) [reply tsl/src/continuous_aggs/invalidation.c:348]: This line is wrong because it should add an entry of `mat_ht`

File: tsl/src/continuous_aggs/invalidation.c
Comment (mfundul) [inline tsl/src/continuous_aggs/invalidation.c:325]: This only gets the status of `raw_ht`, when it should get the status of `mat_ht` for materialized hypertables.

**Review by mfundul — CHANGES_REQUESTED** [review]

File: tsl/src/continuous_aggs/invalidation.c
Comment (fabriziomello) [reply tsl/src/continuous_aggs/invalidation.c:327]: The old code always get the status of this first parameter that is always the `RawHypertable`. If you look below I force the status to by `HypertableIsMaterialization` when we pass the `mat_ht` parameter. The old code path does exactly it but by passing down the status and entry_id (can be the id of `mat_ht` or the `raw_ht`).

And if this new code path is really broken then our regression tests is also broken, no?

Comment (mfundul) [reply tsl/src/continuous_aggs/invalidation.c:327]: That is correct, the regression tests are also broken.
You only check `(mat_ht)` for the distributed case, in the normal case you never go to `case HypertableIsMaterialization:`.

If you look at line `src/process_utility.c:1035`:
```
						ts_cm_functions
							->continuous_agg_invalidate_with_materialization(raw_ht,
																			 mat_ht,
																			 TS_TIME_NOBEGIN,
																			 TS_TIME_NOEND);
```
This will never work for normal hypertables.

**Review by mfundul — APPROVED** [review]
Maybe wait for @mkindahl 's review before merging?

**Comment by fabriziomello** [comment]
> Maybe wait for @mkindahl 's review before merging?

No complains about it and make sense since he wrote the previous code.

File: tsl/src/continuous_aggs/invalidation.c
Comment (svenklemm) [inline tsl/src/continuous_aggs/invalidation.c:361]: this can not happen since it would hit one of the 2 cases above it

Comment (fabriziomello) [reply tsl/src/continuous_aggs/invalidation.c:361]: We build using the `-Werror=switch` compiler option so we need to test all enumeration values, no?

File: tsl/src/continuous_aggs/invalidation.c
Comment (mkindahl) [inline tsl/src/continuous_aggs/invalidation.c:320]: It is odd that you split it up into two functions, but then you put all the logic in one of the functions and use `NULL` indicate the non-materialized table case.

I would have expected the "remove invalidation" function to be split up as well and used, e.g., something like this:

```
void
continuous_agg_invalidate_raw_ht(const Hypertable *raw_ht, int64 start, int64 end)
{
  if (hypertable_is_distributed(raw_ht)) {
    remote_invalidation_log_add_entry(INVALIDATION_HYPER_LOG_ADD_ENTRY_FUNCNAME, raw_ht->fd.id, start, end);
  } else {
    invalidation_hyper_log_add_entry(raw_ht->fd.id, start, end); 
  }
}
```

File: tsl/src/continuous_aggs/invalidation.c
Comment (mkindahl) [inline tsl/src/continuous_aggs/invalidation.c:328]: Similarly, here I would have expected something like:

```
void
continuous_agg_invalidate_mat_ht(const Hypertable *mat_ht, int64 start, int64 end, bool is_distributed)
{
  if (is_distributed) {
    remote_invalidation_log_add_entry(INVALIDATION_CAGG_LOG_ADD_ENTRY_FUNCNAME, mat_ht->fd.id, start, end);
  } else {
    invalidation_hyper_log_add_entry(mat_ht->fd.id, start, end); 
  }
}
```

**Review by mkindahl — COMMENTED** [review]
Why do you disable commit count? This is a single commit.

**Comment by fabriziomello** [comment]
> Why do you disable commit count? This is a single commit.

Removed.

**Review by mkindahl — APPROVED** [review]

**Pull request merged**
